src/classes/RD2_OpportunityService.cls

```diff
@@ -0,0 +1,210 @@
+
+        oppsToInsert = getInstallmentOpportunities(rdsWithCloseDates, contactMap);
+
+        if (oppsToInsert.size() > 0) {
+            for (Database.SaveResult s : insertOpportunities(oppsToInsert)) {
+                if (s.isSuccess() == true) {
+                    //get all the new OppIds into a list
+                    createdOppIds.add(s.getId());
+
+                } else {//handle errors on insert
+                    badSRs.add(s);
+                    badOpps.add(oppsToInsert[oppItem]);
+                }
+                oppItem++;
+            }
+            //Confirm if the error handling is needed due to bulk operations
+            if (!badSRs.isEmpty()) {
+               ERR_Handler.Errors errors = ERR_Handler.getErrors(badSRs, badOpps);
+               ERR_Handler.processErrors(errors.errorRecords, ERR_Handler_API.Context.RD.name());
+            }
+        }
+        updatePaymentRecordOnCreatedOpp(createdOppIds, rdPaymentMethod);
+    }
+
+    /***
+    * @description Gets one or more Opportunity records for a specified set of Recurring Donations with close dates
+    * @param rdsWithCloseDates the Map of Recurring Donations with close dates
+    * @param contactMap the Map of contact with account ids.
+    * @return the list of Opportunities to be created.
+    */
+    private List<Opportunity> getInstallmentOpportunities(Map<npe03__Recurring_Donation__c, List<Date>> rdsWithCloseDates, Map<Id,Contact> contactMap) {
+
+        List<Opportunity> oppsToInsert = new List<Opportunity>();
+        npe03__Recurring_Donations_Settings__c rds = UTIL_CustomSettingsFacade.getRecurringDonationsSettings();
+
+        for (npe03__Recurring_Donation__c r : rdsWithCloseDates.keySet()) {
+
+            for(Integer i = 0;i< rdsWithCloseDates.get(r).size();i++) {
+                Opportunity opp = new Opportunity();
+                opp.Name = r.npe03__Donor_Name__c + ' ' + r.npe03__Amount__c.format() + ' ' + rdsWithCloseDates.get(r)[i].format() + ' ' + System.Label.stgNavRecurringDonations;
+
+                if ( r.npe03__Organization__c != null && String.isBlank(r.npe03__Contact__c)) {
+                    opp.AccountId = r.npe03__Organization__c;
+
+                } else if (contactMap.containsKey(r.npe03__Contact__c)) {
+                    Contact c = contactMap.get(r.npe03__Contact__c);
+                    opp.AccountId = c.AccountId;
+                    opp.Primary_Contact__c = c.id;
+                }
+                opp.CloseDate = rdsWithCloseDates.get(r)[i];
+                opp.StageName = rds.InstallmentOppStageName__c ;
+                opp.Amount = r.npe03__Amount__c;
+                opp.npe03__Recurring_Donation__c = r.Id;
+                if (r.npe03__Recurring_Donation_Campaign__c != null && rds.npe03__Add_Campaign_to_All_Opportunites__c ) {
+                    opp.CampaignId = r.npe03__Recurring_Donation_Campaign__c;
+                }
+                oppsToInsert.add(opp);
```

Comment:
There's one more Opp field we should populate. It's imperfect for now and will require some slight refactoring when the WI to handle Opp naming is done, but for the moment it can be:
```
opp.Recurring_Donation_Installment_Number__c = (rd.npe03__Total_Paid_Installments__c != null ?
    npe03__Total_Paid_Installments__c + 1 : 1);
```

You may need to query this new field when querying the RD records.